exec/Chainweb/Gaps.hs

```diff
@@ -52,79 +46,54 @@ gaps env args = do
       logg Info $ fromString $ show e
     Right cutBS -> gapsCut env args cutBS
 
--- TODO Make this configurable?
-blockQueueSize :: Natural
-blockQueueSize = 30
-
 gapsCut :: Env -> GapArgs -> ByteString -> IO ()
 gapsCut env args cutBS = do
-  let curHeight = fromIntegral $ cutMaxHeight cutBS
-      cids = atBlockHeight curHeight $ _env_chainsAtHeight env
-  getBlockGaps env >>= \gapsByChain ->
+  minHeights <- getAndVerifyMinHeights env cutBS
+  getBlockGaps env minHeights >>= \gapsByChain ->
     if null gapsByChain
       then do
         logg Info $ fromString $ printf "No gaps detected."
         logg Info $ fromString $ printf "Either the database is empty or there are truly no gaps!"
       else do
-        minHeights <- withDb env chainMinHeights
-        when (M.size minHeights /= length cids) $ do
-          logg Error $ fromString $ printf "%d chains have block data, but we expected %d." (M.size minHeights) (length cids)
-          logg Error $ fromString $ printf "Please run a 'listen' first, and ensure that each chain has a least one block."
-          logg Error $ fromString $ printf "That should take about a minute, after which you can rerun 'gaps' separately."
-          exitFailure
         count <- newIORef 0
-        blockQueue <- newTBQueueIO blockQueueSize
-        let strat = maybe Seq (const Par') delay
-            gapSize (a,b) = case compare a b of
+        let gapSize (a,b) = case compare a b of
               LT -> b - a - 1
               _ -> 0
             isGap = (> 0) . gapSize
             total = sum $ fmap (sum . map (bool 0 1 . isGap)) gapsByChain :: Int
-            totalNumBlocks = fromIntegral $ sum $ fmap (sum . map gapSize) gapsByChain :: Int
+            totalNumBlocks = fromIntegral $ sum $ fmap (sum . map gapSize) gapsByChain
         logg Info $ fromString $ printf "Filling %d gaps and %d blocks" total totalNumBlocks
         logg Debug $ fromString $ printf "Gaps to fill %s" (show gapsByChain)
+        let doChain (cid, gs) = do
+              let ranges = concatMap (createRanges cid) gs
+              mapM_ (f logg count cid) ranges
         let gapFiller = do
-              race_ (progress logg count totalNumBlocks) $
-                traverseMapConcurrently_ Par' (\cid -> traverseConcurrently_ strat (f logg blockQueue count cid) . concatMap (createRanges cid)) gapsByChain
+              race_ (progress logg count totalNumBlocks)
+                    (traverseConcurrently_ Par' doChain (M.toList gapsByChain))
```

Comment:
This is the meat of the change.  20 threads (one per chain) instead of 80,000 threads.